ctapipe/io/eventfilereader.py

```diff
@@ -2,36 +2,83 @@
 Handles reading of different event/waveform containing files
 """
 from abc import abstractmethod
-from os.path import basename, splitext, dirname, join, exists
-import numpy as np
-from traitlets import Unicode, Int, CaselessStrEnum, observe
+from os.path import exists
+from traitlets import Unicode, Int, CaselessStrEnum
 from copy import deepcopy
 from ctapipe.core import Component, Factory
 from ctapipe.utils import get_dataset
-from ctapipe.io.hessio import hessio_event_source, hessio_get_list_event_ids
+from ctapipe.core import Provenance
 
 
 class EventFileReader(Component):
     """
-    Parent class for EventFileReaders of different sources. A new
-    EventFileReader should be created for each type of event file read into
-    ctapipe, e.g. simtelarray files are read by the `HessioFileReader`.
+    Parent class for EventFileReaders of different sources.
+
+    A new EventFileReader should be created for each type of event file read
+    into ctapipe, e.g. sim_telarray files are read by the `HessioFileReader`.
+
+    EventFileReader provides a common high-level interface for accessing event
+    data from different data sources. Creating an EventFileReader for a new
+    data source ensures that data can be accessed in a common way,
+    irregardless of the data source.
+
+    EventFileReader itself is an abstract class. To use an EventFileReader you
+    must use a subclass that is relevant for the data format for the file you
+    are reader (for example you must use
+    `ctapipe.io.hessiofilereader.HessioFileReader` to read a hessio format
+    file). Alternatively you can use
+    `ctapipe.io.eventfilereader.EventFileReaderFactory` to automatically
+    select the correct EventFileReader subclass for the data format you wish
+    to read.
+
+    To loop through the events in a file:
+
+    >>> reader = EventFileReader(None, None, input_path=path)
+    >>> for event in reader:
+    >>>    print(event.count)
+
+    **NOTE**: Every time a new loop is started through the reader, it restarts
+    from the first event.
+
+    To obtain a particular event in a file:
+
+    >>> reader = EventFileReader(None, None, input_path=path)
+    >>> event = reader[event_index]
+    >>> print(event.count)
+
+    To obtain a particular event in a file from its event_id:
+
+    >>> reader = EventFileReader(None, None, input_path=path)
+    >>> event = reader["event_id"]
```

Comment:
Well, at least in sim_telarray, event_id does not increase in increments of 1. I think this is because only shower events that triggered the array are stored, but sim_telarray increments the id on all events (take a look at event.r0.event_id for a hessio file).

Having event_index allows the user to not care about this, and just get the next event.

But perhaps the user has noticed something peculiar about a particular event, perhaps in a different analysis software, allowing them to obtain an event using event_id allows them to ensure the same event is obtained between the programs.

Reply:
> Well, at least in sim_telarray, event_id does not increase in increments of 1. 

Ah yes, this is indeed quite usual in IACT simulation. First N air showers are simulated, the M < N events are triggered and stored in a file. 
One might think events could be numbered consecutively, but for repeatability, the random seeds and all settings are stored somewhere, and in order to reproduce a single event later, one needs to know the "airshower-id" it is associated with and not the "trigger-id".

However this pattern goes on, right? After that, some kind of cleaning is applied and maybe a cut, which cuts away all events that did not "survive cleaning". 

So for simulated events, each event now has: "airshower-id", "trigger-id", "survived-cleaning-id" 

And we can go on: "survived-some-precuts", "survived-background-cuts", "survived-some-quality-cuts" ...

Any process, which selects some events out of many other events, results in a new "sub-stream", where one might say: 

"Hery! Why are the ids of these events not consecutively increasing but have "holes" in them?"

I do not think that the solution is to invent new beautiful ids after each cut, but to accept that event ids are simply not increasing in steps of 1. They are not so much counters but "telephone-numbers" each event has one .. it is unique ... but it is not increasing in steps of 1.

I would discourage people from inventing two different but unique ids for events. Since, how do people talk to each other then? 

"Hey have you looked at event 14532 in run 20190101_1432?" 

"what do you mean id=14532 or index=14532?"

"erm .. wait a sec ... dunno I closed the window"

Not so useful, is it?

Reply:
Sure, but theres at least 2 clear identification of events:
event_id: The ID associated by the software or camera (or CTA trigger in the final array). So can be thought of an external ID unassociated with ctapipe.
event_index: The ID associated with the event by ctapipe, assigned in the order the events are read from the file.

I think this is clear enough. I don't think we can rely on event_id alone. 
1. Some sim_telarray files are combined with others into a final file, allowing for duplicate event_ids.
2. With `__gettitem__` I wish to be able to get the first or the nth event in the file, a system that doesn't work if I cannot get an event using event_index (i.e. order they are read from the file).
3. Theres nothing to say event_id 101 is directly after event_id 100. So we can't do anything smart in `__getitem__` to avoid looping from the start of the file again to find it. But event_index 4 is directly after event_index 3, so the loop to find an event based on event_index can operate smarter and just continue onto the next event in the generator without starting from the beginning in the file.

Reply:
There's a CTA requirement that we have a unique ID for each event, but so far the MCs do not support that (though we could somehow combine the `mc_shower_id` and the `mc_event_id` into one unique identifier).  They will probably never increment by 1, since you may lose events (non-triggered, didn't read out due to dead time) at the DAQ level, or during data volume reduction.